include/nil/crypto3/detail/pack.hpp

```diff
@@ -1,232 +1,864 @@
+
+#elif defined(CRYPTO3_TARGET_CPU_IS_BIG_ENDIAN)
+            template<int UnitBits, int ValueBits, typename InT, typename OutT>
+            struct can_memcpy<stream_endian::big_unit_big_bit<UnitBits>, ValueBits, InT, OutT>
+                : host_can_memcpy<UnitBits, ValueBits, InT, OutT> { };
             template<int UnitBits, int InputBits, int OutputBits, typename InT, typename OutT>
-            struct can_memcpy<stream_endian::little_unit_big_bit<UnitBits>, InputBits, OutputBits, InT, OutT>
-                : host_can_memcpy<UnitBits, InputBits, OutputBits, InT, OutT> {};
-            template<int UnitBits, int InputBits, int OutputBits, typename InT, typename OutT>
-            struct can_memcpy<stream_endian::little_unit_little_bit<UnitBits>, InputBits, OutputBits, InT, OutT>
-                : host_can_memcpy<UnitBits, InputBits, OutputBits, InT, OutT> {};
+            struct can_memcpy<stream_endian::big_unit_little_bit<UnitBits>, ValueBits, InT, OutT>
+                : host_can_memcpy<UnitBits, ValueBits, InT, OutT> { };
 #endif
 
-#ifdef CRYPTO3_TARGET_CPU_IS_BIG_ENDIAN
-            template<int UnitBits, int InputBits, int OutputBits, typename InT, typename OutT>
-            struct can_memcpy<stream_endian::big_unit_big_bit<UnitBits>, InputBits, OutputBits, InT, OutT>
-                : host_can_memcpy<UnitBits, InputBits, OutputBits, InT, OutT> {};
-            template<int UnitBits, int InputBits, int OutputBits, typename InT, typename OutT>
-            struct can_memcpy<stream_endian::big_unit_little_bit<UnitBits>, InputBits, OutputBits, InT, OutT>
-                : host_can_memcpy<UnitBits, InputBits, OutputBits, InT, OutT> {};
-#endif
+            /*!
+             * @brief Packer is used to transform input data divided into chunks of the bit size InputValueBits
+             * represented in input endianness (InputEndianness)
+             * into output data (of the same bit length) divided into chunks of the bit size OutputValueBits
+             * represented in output endianness (OutputEndianness).
+             *
+             * The choice of packer depends on the following conditions:
+             * 1. input and output chunk size relation (equal, less, or greater);
+             * 2. input and output endianness relation (same or different);
+             * 3. the possibility of fast data copy using memcpy.
+             *
+             * @ingroup pack
+             *
+             * @tparam InputEndianness
+             * @tparam OutputEndianness
+             * @tparam InputValueBits
+             * @tparam OutputValueBits
+             * @tparam InputType
+             * @tparam OutputType             
+             * @tparam SameEndianness
+             * @tparam Implode
+             * @tparam Explode
+             */
+            template<typename InputEndianness, typename OutputEndianness, std::size_t InputValueBits,
+                     std::size_t OutputValueBits, typename InputType, typename OutputType,
```

Comment:
I'm pretty sure this should be factored out to the separate parent class, which will be required to be responsible for the selection of particular algorithm. Partial specializations for ```InputValue``` and ```OutputValue``` would allow to implement ```bool```-specific algorithms.

Reply:
Parent class-selector is written, all unnecessary static asserts removed.